[SPARK-29323][WEBUI] Add tooltip for The Executors Tab's column names in the Spark history server Page

liucht-inspur opened the pull request:
### What changes were proposed in this pull request?

This PR is Adding tooltip for The Executors Tab's column names include RDD Blocks, Disk Used,Cores, Activity Tasks, Failed Tasks , Complete Tasks, Total Tasks in the history server Page.
https://issues.apache.org/jira/browse/SPARK-29323

![image](https://user-images.githubusercontent.com/28332082/66017759-b6c24a80-e50e-11e9-807b-5b076f701d2f.png)

I have modify the following code in executorspage-template.html
Before:
<th>RDD Blocks</th>
<th>Disk Used</th>
<th>Cores</th>
<th>Active Tasks</th>
<th>Failed Tasks</th>
<th>Complete Tasks</th>
<th>Total Tasks</th>

![image](https://user-images.githubusercontent.com/28332082/66018111-4ddbd200-e510-11e9-9cfc-19f3eae81e76.png)

After:

<th><span data-toggle="tooltip" data-placement="top" title="RDD Blocks">RDD Blocks</span></th>
<th><span data-toggle="tooltip" data-placement="top" title="Disk Used">Disk Used</span></th>
<th><span data-toggle="tooltip" data-placement="top" title="Cores">Cores</span></th>
<th><span data-toggle="tooltip" data-placement="top" title="Active Tasks">Active Tasks</span></th>
<th><span data-toggle="tooltip" data-placement="top" title="Failed Tasks">Failed Tasks</span></th>
<th><span data-toggle="tooltip" data-placement="top" title="Complete Tasks">Complete Tasks</span></th>
<th><span data-toggle="tooltip" data-placement="top" title="Total Tasks">Total Tasks</span></th>
   

![image](https://user-images.githubusercontent.com/28332082/66018157-79f75300-e510-11e9-96ba-6230aa0940c7.png)
    

### Why are the changes needed?

the spark Executors of history Tab page, the Summary part shows the line in the list of title, but format is irregular.
Some column names have tooltip, such as Storage Memory, Task Time(GC Time), Input, Shuffle Read,
Shuffle Write and Blacklisted, but there are still some list names that have not tooltip. They are RDD Blocks, Disk Used,Cores, Activity Tasks, Failed Tasks , Complete Tasks and Total Tasks. oddly, Executors section below,All the column names Contains the column names above have tooltip .
It's important for open source projects to have consistent style and user-friendly UI, and I'm working on keeping it consistent And more user-friendly.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Manual tests for Chrome, Firefox and Safari


Authored-by: liucht-inspur <liucht@inspur.com>
Signed-off-by: liucht-inspur <liucht@inspur.com>

liucht-inspur committed:
This PR is Adding tooltip for The Executors Tab's column names include
RDD Blocks, Disk Used,Cores, Activity Tasks, Failed Tasks , Complete
Tasks, Total Tasks in the history server Page.

Before:

<th>RDD Blocks</th>
<th>Disk Used</th>
<th>Cores</th>
<th>Active Tasks</th>
<th>Failed Tasks</th>
<th>Complete Tasks</th>
<th>Total Tasks</th>

After:

<th><span data-toggle="tooltip" data-placement="top" title="RDD
Blocks">RDD Blocks</span></th>
<th><span data-toggle="tooltip" data-placement="top" title="Disk
Used">Disk Used</span></th>
<th><span data-toggle="tooltip" data-placement="top"
title="Cores">Cores</span></th>
<th><span data-toggle="tooltip" data-placement="top" title="Active
Tasks">Active Tasks</span></th>
<th><span data-toggle="tooltip" data-placement="top" title="Failed
Tasks">Failed Tasks</span></th>
<th><span data-toggle="tooltip" data-placement="top" title="Complete
Tasks">Complete Tasks</span></th>
<th><span data-toggle="tooltip" data-placement="top" title="Total
Tasks">Total Tasks</span></th>

	### Why are the changes needed?

	the spark Executors of history Tab page, the Summary part shows
	the line in the list of title, but format is irregular.
	Some column names have tooltip, such as Storage Memory, Task
	Time(GC Time), Input, Shuffle Read,
	Shuffle Write and Blacklisted, but there are still some list
	names that have not tooltip. They are RDD Blocks, Disk
	Used,Cores, Activity Tasks, Failed Tasks , Complete Tasks and
	Total Tasks. oddly, Executors section below,All the column names
	Contains the column names above have tooltip .
	It's important for open source projects to have consistent style
	and user-friendly UI, and I'm working on keeping it consistent
	And more user-friendly.

	### Does this PR introduce any user-facing change?

	No.

	### How was this patch tested?

	Manual tests for Chrome, Firefox and Safari

        Authored-by: liucht-inspur <liucht@inspur.com>
        Signed-off-by: liucht-inspur <liucht@inspur.com>

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

liucht-inspur reviewed (commented):
Please help to review this change,thanks.

srowen commented:
Where does the tooltip text come from? I'm missing that part.

liucht-inspur commented:
@srowen They come from the same column names in the Executors like that under the red line: 
![image](https://user-images.githubusercontent.com/28332082/66135394-778a1b80-e62c-11e9-8c9d-df9d79c80c7b.png)

liucht-inspur commented:
Summary code as

![image](https://user-images.githubusercontent.com/28332082/66137252-6262bc00-e62f-11e9-8694-2cbfcb049b38.png)


PK Executors code as

![image](https://user-images.githubusercontent.com/28332082/66137217-5545cd00-e62f-11e9-8b9c-1da38d0876b4.png)
![image](https://user-images.githubusercontent.com/28332082/66137466-b5d50a00-e62f-11e9-9c2b-a5f163f2f542.png)

liucht-inspur commented:
> Where does the tooltip text come from? I'm missing that part.

Like the Executors' column names

srowen commented:
I mean, what text would the tooltip now show for the additional columns?

liucht-inspur commented:
> I mean, what text would the tooltip now show for the additional columns?

I copy the code from Executors lable's column tooltip ，Before I revised  it, it is like this

Before modification
 Executors code as

![image](https://user-images.githubusercontent.com/28332082/66137217-5545cd00-e62f-11e9-8b9c-1da38d0876b4.png)
![image](https://user-images.githubusercontent.com/28332082/66137466-b5d50a00-e62f-11e9-9c2b-a5f163f2f542.png)

srowen commented:
I don't think that answers the question. What is the new tooltip that appears on what columns?

liucht-inspur commented:
> Can one of the admins verify this patch?



> I don't think that answers the question. What is the new tooltip that appears on what columns?

Hello @srowen , I had screenshots and added comments, hope they are valuable to you,thanks!

![image](https://user-images.githubusercontent.com/28332082/66178256-485cc400-e697-11e9-93e2-e582bdda0910.png)

liucht-inspur commented:
@srowen My change is that 
the column **RDD Blocks** add tooltip text 'RDD Blocks',
the column **Disk Used** add tooltip text 'Disk Used',
the column **Cores** add tooltip text 'Cores',
the column **Active Tasks** add tooltip text 'Active Tasks',
the column **Failed Tasks** add tooltip text 'Failed Tasks',
the column **Complete Tasks** add tooltip text 'Complete Tasks',
the column **Total Tasks** add tooltip text 'Total Tasks',

srowen commented:
What is the point of the tooltip then? it just repeats the column name.

liucht-inspur commented:
> What is the point of the tooltip then? it just repeats the column name.

@srowen  It's important for open source projects to have consistent style and user-friendly UI, and I'm working on keeping it consistent And more user-friendly.

srowen commented:
Hm, no I don't think that adds anything to user-friendliness. It's just a redundant popup. If you made the tooltips meaningful, maybe so.

liucht-inspur commented:
So, can you give some suggestions about the tootip text?

liucht-inspur commented:
> Hm, no I don't think that adds anything to user-friendliness. It's just a redundant popup. If you made the tooltips meaningful, maybe so.

@srowen I think it's necessary to be consistent in any case.

srowen commented:
I don't think there's much meaningful to say in a tooltip here, so no I just wouldn't make this change.

liucht-inspur commented:
@srowen For the same problem, I suggest adding more information in the columns of tooltip Active Tasks and Failed Tasks, because these two columns have different background colors, such as **blue**, **red**, and **white**, and they have **different transparency**, which is really confusing, what do you think?
See the columns of different background colors shown in the figure below
![image](https://user-images.githubusercontent.com/28332082/66372968-fc34bb00-e9d9-11e9-97df-934c1a18bb33.png)

srowen commented:
Yes, explaining the colors would be a reasonable role for a tooltip.

liucht-inspur committed:
add more infomation for active task and failed task to explaining the colors and color opacity

liucht-inspur commented:
@srowen ok,thank you ! I have finished this change,

two changes:
1、in Summary 
before:
`<th>Active` `Tasks</th>`
`<th>Failed` `Tasks</th>`

after:
`<th><span` data-toggle="tooltip" data-placement="top" title="Active Tasks show in shaded blue. % of color opacity range from 0 to `maxTasks.">Active` `Tasks</span></th>`
`<th><span` data-toggle="tooltip" data-placement="top" title="Failed Tasks show in shaded red. Color opacity reach max at 10% of Total Tasks.">Failed `Tasks</span></th>`

added tooltip for Active Tasks and Failed Tasks ,the titles are 
"Active Tasks show in shaded blue. % of color opacity range from 0 to maxTasks." and "Failed Tasks show in shaded red. Color opacity reach max at 10% of Total Tasks."

2、in Executors
before:
`<th><span` data-toggle="tooltip" data-placement="top" title="Active Tasks">Active `Tasks</span></th>`
`<th><span` data-toggle="tooltip" data-placement="top" title="Failed Tasks">Failed `Tasks</span></th>`

after:
`<th><span` data-toggle="tooltip" data-placement="top" title="Active Tasks show in shaded blue. % of color opacity range from 0 to maxTasks.">Active `Tasks</span></th>`
`<th><span` data-toggle="tooltip" data-placement="top" title="Failed Tasks show in shaded red. Color opacity reach max at 10% of Total Tasks.">Failed `Tasks</span></th>`

added more information for tooltip

Please help to review this change,thanks.

srowen commented on core/src/main/resources/org/apache/spark/ui/static/executorspage-template.html:
```diff
@@ -37,8 +37,8 @@ <h4 class="title-table">Summary</h4>
-        <th>Active Tasks</th>
-        <th>Failed Tasks</th>
+        <th><span data-toggle="tooltip" data-placement="top" title="Active Tasks show in shaded blue. % of color opacity range from 0 to maxTasks.">Active Tasks</span></th>
+        <th><span data-toggle="tooltip" data-placement="top" title="Failed Tasks show in shaded red. Color opacity reach max at 10% of Total Tasks.">Failed Tasks</span></th>
```
Likewise, maybe:
"Number of tasks that have failed on this executor. Darker shading highlights executors with a high proportion of failed tasks."

liucht-inspur replied:
Yes, they are more suitable，thank you！

liucht-inspur committed:
modify the infomation for tooltips of Active Tasks and Failed Tasks

liucht-inspur commented:
@srowen Thanks again, I have finished the change.

srowen commented:
We'll have to wait for Jenkins to come back, to evaluate this with tests, but I'm sure it will be OK.

liucht-inspur commented:
cc @wangyum

wangyum commented:
Sorry @liucht-inspur This has nothing to do with authorization:
http://apache-spark-developers-list.1001551.n3.nabble.com/build-system-IMPORTANT-northern-california-fire-danger-potential-power-outage-s-td28059.html

liucht-inspur commented:
Ok, I see. We will wait quietly. Thank you all!

wangyum commented:
retest this please

SparkQA commented:
**[Test build #111957 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111957/testReport)** for PR 25994 at commit [`49ffc0c`](https://github.com/apache/spark/commit/49ffc0c9172810ce65ba05e556323cadd83bc98b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16973/
Test PASSed.

SparkQA commented:
**[Test build #111957 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111957/testReport)** for PR 25994 at commit [`49ffc0c`](https://github.com/apache/spark/commit/49ffc0c9172810ce65ba05e556323cadd83bc98b).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111957/
Test FAILed.

wangyum commented:
retest this please

SparkQA commented:
**[Test build #111968 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111968/testReport)** for PR 25994 at commit [`49ffc0c`](https://github.com/apache/spark/commit/49ffc0c9172810ce65ba05e556323cadd83bc98b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16983/
Test PASSed.

SparkQA commented:
**[Test build #111968 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111968/testReport)** for PR 25994 at commit [`49ffc0c`](https://github.com/apache/spark/commit/49ffc0c9172810ce65ba05e556323cadd83bc98b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111968/
Test PASSed.

srowen commented:
Merged to master

AbhishekNew commented:
@liucht-inspur Did u handle the tooltip for the Live UI Page for Spark.

liucht-inspur commented:
> @liucht-inspur Did u handle the tooltip for the Live UI Page for Spark.

Yeah, exactly
